build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.build-logic.groovy-dsl-gradle-plugin.gradle.kts

```diff
@@ -50,11 +50,23 @@ tasks.withType<GroovyCompile>().configureEach {
 }
 
 tasks.withType<Test>().configureEach {
-    if (JavaVersion.current().isJava9Compatible) {
+    val testVersionProvider = javaLauncher.map { it.metadata.languageVersion }
+    jvmArgumentProviders.add(CommandLineArgumentProvider {
         //allow ProjectBuilder to inject legacy types into the system classloader
-        jvmArgs("--add-opens", "java.base/java.lang=ALL-UNNAMED")
-        jvmArgs("--illegal-access=deny")
-    }
+        if (testVersionProvider.get().canCompileOrRun(9)) {
+            listOf("--add-opens", "java.base/java.lang=ALL-UNNAMED")
+        } else {
+            emptyList()
+        }
+    })
+    jvmArgumentProviders.add(CommandLineArgumentProvider {
+        val testVersion = testVersionProvider.get()
+        if (testVersion.canCompileOrRun(9) && !testVersion.canCompileOrRun(17)) {
```

Comment:
We only want to add `--illegal-access=deny` on versions older than 17, as in Java 17 that became the default. In versions after 17, it is an error to include this flag